typescript/token/contracts/libs/FastTokenRouter.sol

```diff
@@ -0,0 +1,195 @@
+        bytes calldata _message
+    ) internal virtual override {
+        bytes32 recipient = _message.recipient();
+        uint256 amount = _message.amount();
+        bytes calldata metadata = _message.metadata();
+        _transferTo(recipient.bytes32ToAddress(), amount, _origin, metadata);
+        emit ReceivedTransferRemote(_origin, recipient, amount);
+    }
+
+    /**
+     * @dev Transfers `_amount` of token to `_recipient`/`fastFiller` who provided LP.
+     * @dev Called by `handle` after message decoding.
+     */
+    function _transferTo(
+        address _recipient,
+        uint256 _amount,
+        uint32 _origin,
+        bytes calldata _metadata
+    ) internal virtual {
+        address _tokenRecipient = _getTokenRecipient(
+            _recipient,
+            _amount,
+            _origin,
+            _metadata
+        );
+
+        _fastTransferTo(_tokenRecipient, _amount);
+    }
+
+    /**
+     * @dev allows an external user to full an unfilled fast transfer order.
+     * @param _recipient The recepient of the wrapped token on base chain.
+     * @param _amount The amount of wrapped tokens that is being bridged.
+     * @param _fastFee The fee the bridging entity will pay.
+     * @param _fastTransferId Id assigned on the remote chain to uniquely identify the transfer.
+     */
+    function fillFastTransfer(
+        address _recipient,
+        uint256 _amount,
+        uint256 _fastFee,
+        uint32 _origin,
+        uint256 _fastTransferId
+    ) external virtual {
+        bytes32 filledFastTransfersKey = keccak256(
+            abi.encodePacked(_origin, _fastTransferId)
+        );
+        require(
+            filledFastTransfers[filledFastTransfersKey].filler == address(0),
+            "request already filled"
+        );
+
+        filledFastTransfers[filledFastTransfersKey] = FastTranferMetadata(
+            msg.sender,
+            _recipient,
+            _amount,
+            _fastFee
+        );
+
+        _fastRecieveFrom(msg.sender, _amount - _fastFee);
+        _fastTransferTo(_recipient, _amount - _fastFee);
+    }
+
+    /**
+     * @dev Transfers `_amountOrId` token to `_recipient` on `_destination` domain.
+     * @dev Delegates transfer logic to `_fastTransferFromSender` implementation.
+     * @dev Emits `SentTransferRemote` event on the origin chain.
+     * @param _destination The identifier of the destination chain.
+     * @param _recipient The address of the recipient on the destination chain.
+     * @param _amountOrId The amount or identifier of tokens to be sent to the remote recipient.
+     * @return messageId The identifier of the dispatched message.
+     */
+    function fastTransferRemote(
+        uint32 _destination,
+        bytes32 _recipient,
+        uint256 _amountOrId,
+        uint256 _fastFee
+    ) public payable virtual returns (bytes32 messageId) {
+        uint256 _fastTransferId = fastTransferId;
+        fastTransferId = _fastTransferId + 1;
```

Comment:
I meant if you have the transfer ID be a commitment to the origin and transfer metadata
```suggestion
        bytes32 fastTransferId = keccak256(fastTransferId, fastTransferMetadata);
```

Reply:
Because the handle does not know the filler address, that can't be hashed.
But right we can probably hash the rest of the data and store that as a key and ensure it uses a single slot for address.

Reply:
hmm good point, I like your change